### PR TITLE
[RFR] Play over the internet

### PR DESCRIPTION
**Pull request by benjaminpayoux**
[Trello](https://trello.com/c/O6I8o0qx/7-05-as-peter-i-would-like-to-be-able-to-play-over-the-internet)

# TODO

- [X] Play from different devices
- [X] Players only play on their turn
- [x] Find a way to refresh when opponent is playing

# Result
last

### CODE REVIEW

File: src/Controller/GameController.php [hunk 1]
```diff
@@ -29,16 +34,31 @@ class GameController extends AbstractController
      */
     private $translator;
 
+    /**
+     * @var PlayerProvider $playerProvider
+     */
+    private $playerProvider;
+    private $session;
```
Comment (jdemangeon) [inline src/Controller/GameController.php]: Missing descriptions for session and playerManipulator

File: src/Controller/GameController.php [hunk 2]
```diff
@@ -50,6 +70,9 @@ public function index(Request $request, Game $game): Response
             return $this->render('abandoned.html.twig');
         }
 
+        $player = $this->playerProvider->getPlayer($game);
```
Comment (jdemangeon) [inline src/Controller/GameController.php]: $gamePlayer is more relevant IMO

File: src/Controller/WelcomeController.php [hunk 3]
```diff
@@ -37,6 +39,8 @@ public function index(Request $request): Response
         if ($form->isSubmitted() && $form->isValid()) {
             $game = $this->gameManipulator->createGame();
 
+            $this->session->set('player_hash', $game->getPlayer1()->getHash());
```
Comment (jdemangeon) [inline src/Controller/WelcomeController.php]: You should manipule player session from your "player management" class.
The "player_hash" need to be in one place only

File: src/GameManipulator.php [hunk 4]
```diff
@@ -203,12 +224,13 @@ public function switchCurrentPlayer($game): Game
      * 
      * @param Shoot $shoot
      * @param Game $game
+     * @param Player $player
      * 
      * @return Ship|null
      */
-    public function didShootHitOpponentShip(Shoot $shoot, Game $game): ?Ship
+    public function didShootHitOpponentShip(Shoot $shoot, Game $game, Player $player): ?Ship
```
Comment (jdemangeon) [inline src/GameManipulator.php]: Better! But i'm still confused about which player it is? The current one? The opponent?
Using another name can fix this (eg: shootingPlayer)

File: src/GameManipulator.php [hunk 5]
```diff
@@ -227,15 +249,16 @@ public function didShootHitOpponentShip(Shoot $shoot, Game $game): ?Ship
      * 
      * @param Ship $ship
      * @param Game $game
+     * @param Player $player
      * 
      * @return bool
      */
-    public function isShipSunk(Ship $ship, Game $game): bool
+    public function isShipSunk(Ship $ship, Game $game, Player $player): bool
```
Comment (jdemangeon) [inline src/GameManipulator.php]: Better! But i'm still confused about which player it is? The current one? The opponent?

File: src/GameManipulator.php [hunk 6]
```diff
@@ -254,108 +277,82 @@ public function isShipSunk(Ship $ship, Game $game): bool
      * Get opponent ships that have been sunk
      * 
      * @param Game $game 
+     * @param Player $player
      *  
      * @return array
      */
-    public function getOpponentShipsSunk(Game $game): array {
+    public function getOpponentShipsSunk(Game $game, Player $player): array
```
Comment (jdemangeon) [inline src/GameManipulator.php]: Better! But i'm still confused about which player it is? The current one? The opponent?

File: src/GameManipulator.php [hunk 7]
```diff
@@ -254,108 +277,82 @@ public function isShipSunk(Ship $ship, Game $game): bool
...
      * 
      * @return Ship
      */
     public function getRandShip(int $length, string $orientation, Game $game): Ship
     {
         $positionOk = false;
 
-        if ($orientation === GameConstants::HORIZONTAL)
-        {
+        if ($orientation === GameConstants::HORIZONTAL) {
```
Comment (jdemangeon) [inline src/GameManipulator.php:314]: You've a linter, good :grin:

File: src/Provider/PlayerProvider.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     * 
+     * @param Game $game
+     * @return Player
+     */
+    public function getPlayer(Game $game): ?Player
+    {
+        $playerHash = $this->session->get('player_hash');
+
+        // For demo purpose, we consider that only the second player comes
+        // TODO: better management of the security
```
Comment (jdemangeon) [inline src/Provider/PlayerProvider.php:59]: Good :+1: It help the reviewer to known your intention

File: src/Repository/ShipRepository.php [hunk 9]
```diff
@@ -44,6 +44,7 @@ public function getCurrentPlayerShips(Game $game): array
      */
     public function getPlayerShips(Game $game, Player $player): array
     {
-        return $this->findBy(['game' => $game, 'player' => $player]);    
+        $ships = $this->findBy(['game' => $game, 'player' => $player]);
```
Comment (jdemangeon) [inline src/Repository/ShipRepository.php]: Direct return will do the trick

**Review by jdemangeon — CHANGES_REQUESTED** [review]
It's far better than the start, you progress at the speed of light. 
I'm sure you're a Symfony fan right now. 

Go home, you drunk Laravel ! :beer:

File: templates/game.html.twig [hunk 10]
```diff
@@ -1,5 +1,12 @@
 {% extends "base.html.twig" %}
 
+{% block head %}
+	{{ parent() }}
+    {% if (ships|first).player.id != game.currentPlayer.id %}
+	    <meta http-equiv="refresh" content="5">
```
Comment (jdemangeon) [inline templates/game.html.twig]: High level engineering :grin:

Comment (jdemangeon) [reply templates/game.html.twig]: Can you set a value to known you intention?

`{% set isOponentTurn = (ships|first).player.id != game.currentPlayer.id %}`

File: src/Repository/ShipRepository.php (hunk 9 above)
Comment (benjaminpayoux) [reply src/Repository/ShipRepository.php]: Oops. Was for debugging

File: templates/game.html.twig (hunk 10 above)
Comment (benjaminpayoux) [reply templates/game.html.twig]: Fixed. I added the game player to the view variable.

**Review by jdemangeon — APPROVED** [review]

**Pull request merged**